python/symbolica_community/tensors.pyi

```diff
@@ -0,0 +1,307 @@
+        Get the flattened index of the given indices.
+        """
+
+class TensorNetwork:
+    r"""
+    A tensor network.
+
+    This class is a wrapper around the `TensorNetwork` class from the `spenso` crate.
+    Such a network is a graph representing the arithmetic operations between tensors.
+    In the most basic case, edges represent the contraction of indices.
+    """
+    def __new__(cls,expr:Expression) -> TensorNetwork:
+        r"""
+        Create a new tensor network.
+        """
+
+    def contract(self):
+        r"""
+        Contract the tensor network.
+        """
+
+    def result(self)->Tensor:
+        r"""
+        Return the result of the contraction. This is only valid after the network has been contracted.
+        """
+
+    def __str__(self) -> str:
+        ...
+
+class TensorStructure:
+    r"""
+    A structure that can be used to represent the "shape" of a tensor.
+    This has an optional name, and accompanying symbolica expressions that are considered as additional non-indexed arguments.
+    The structure is essentially a list of `Representation` that are used to define the structure of the tensor.
+    """
+
+    def __new__(cls,*additional_args:Representation|Expression,name:Optional[Expression]=None) -> TensorStructure:
+        r"""
+        Create a new tensor structure with the given name and representations. Any additional expressions are also passed along.
+        """
+
+    def __repr__(self) -> str:
+        ...
+
+    def __str__(self) -> str:
+        ...
+
+    def __len__(self) -> int:
+        r"""
+        Return the size of the tensor.
+        """
+
+    @overload
+    def __getitem__(self, idx: int) -> List[int]:
+        r"""
+        Get the indices of the slot at the given flat index.
+        """
+
+    @overload
+    def __getitem__(self, s: slice) -> Sequence[List[int]]:
+        r"""
+        Get the list of expanded indices of the slot at the given slice of flat index.
+        """
+
+    @overload
+    def __getitem__(self, key: List[int]) -> int:
+        r"""
+        Get the flattened index of the given indices.
+        """
+
+
+
+def register(tensor:Tensor):
+    r"""
+    Register the tensor in the tensor registry. The tensor must have at least a name.
+    """
+
+def dense(structure:List[int]|List[Slot]|List[Representation]|TensorIndices|TensorStructure,data:List[float]|List[Expression]) -> Tensor:
+    r"""
+    Create a new dense tensor with the given structure and data.
+    The structure can be a list of integers, a list of representations, or a list of slots.
+    In the first two cases, no "indices" are assumed, and thus the tensor is indexless (i.e.) it has a shape but no proper way to contract it.
+    The structure can also be a proper `TensorIndices` object or `TensorStructure` object.
+
+    The data is either a list of floats or a list of symbolica expressions, of length equal to the number of elements in the structure, in row-major order.
+    """
+    ...
```

Comment:
What are these triple dots?

Reply:
They were added automatically by the pyo3 stub gen. But I think they just explicitly show that this is a stub (i.e. that the body isn't there)